plugin/src/main/java/org/opensearch/ml/action/models/DeleteModelTransportAction.java

```diff
@@ -234,6 +259,126 @@ public void onFailure(Exception e) {
+            if (allDependentPipelineIds.isEmpty()) {
+                actionListener.onResponse(true);
+            } else {
+                actionListener
+                    .onFailure(
+                        new OpenSearchStatusException(
+                            String
+                                .format(
+                                    Locale.ROOT,
+                                    "%d ingest pipelines are still using this model, please delete or update the pipelines first: %s",
+                                    allDependentPipelineIds.size(),
+                                    Arrays.toString(allDependentPipelineIds.toArray(new String[0]))
+                                ),
+                            RestStatus.CONFLICT
+                        )
+                    );
+            }
+        }, e -> {
+            log.error("Failed to delete ML Model: " + modelId, e);
+            actionListener.onFailure(e);
+
+        }));
+
+    }
+
+    private void checkSearchPipelineBeforeDeleteModel(String modelId, ActionListener<Boolean> actionListener) {
+        GetSearchPipelineRequest getSearchPipelineRequest = new GetSearchPipelineRequest();
+        client.execute(GetSearchPipelineAction.INSTANCE, getSearchPipelineRequest, ActionListener.wrap(searchPipelineResponse -> {
+            List<String> allDependentPipelineIds = findDependentPipelines(
+                searchPipelineResponse.pipelines(),
+                modelId,
+                org.opensearch.search.pipeline.PipelineConfiguration::getConfigAsMap,
+                org.opensearch.search.pipeline.PipelineConfiguration::getId
+            );
+            if (allDependentPipelineIds.isEmpty()) {
+                actionListener.onResponse(true);
+            } else {
+                actionListener
+                    .onFailure(
+                        new OpenSearchStatusException(
+                            String
+                                .format(
+                                    Locale.ROOT,
+                                    "%d search pipelines are still using this model, please delete or update the pipelines first: %s",
+                                    allDependentPipelineIds.size(),
+                                    Arrays.toString(allDependentPipelineIds.toArray(new String[0]))
+                                ),
+                            RestStatus.CONFLICT
+                        )
+                    );
+            }
+        }, e -> {
+            log.error("Failed to delete ML Model: " + modelId, e);
+            actionListener.onFailure(e);
+
+        }));
+
+    }
+
+    private void checkDownstreamTaskBeforeDeleteModel(String modelId, Boolean isHidden, ActionListener<DeleteResponse> actionListener) {
+        CountDownLatch countDownLatch = new CountDownLatch(3);
+        AtomicBoolean noneBlocked = new AtomicBoolean(true);
+        CopyOnWriteArrayList<String> errorMessages = new CopyOnWriteArrayList<>();
```

Comment:
It's good to use a thread safe vector here, but a suggestion is to change to ConcurrentLinkedQueue, this has better performance than CopyOnWriteArrayList as it doesn't need to copy the whole list when inserting new elements, and internally it's based on CAS so I believe it has best performance among all other thread safe list/queue.

Reply:
Great. Already change it.